### PR TITLE
[1.x] schema for channles

### PR DESCRIPTION
**Pull request by FredPeal**

### CODE REVIEW

File: graphql/schemas/Social/channels.graphql [hunk 1]
```diff
@@ -0,0 +1,46 @@
+type SocialChannel {
+    id: ID!
+    name: String!
+    slug: String
+    description: String!
+    entity_namespace: String!
+    entity_id: ID!
+    last_message_id: ID!
+    messages: [Messages]
```
Comment (kaioken) [inline graphql/schemas/Social/channels.graphql:9]: @FredPeal list have to follow the standard [Message!]! @hasMany

File: graphql/schemas/Social/message.graphql [hunk 2]
```diff
@@ -14,13 +14,27 @@ type Message {
...
+enum DistributionType {
+    ALL @enum(value: 1)
+    Channels: @enum(value: 2)
+    Followers: @enum(value: 3)
+}
+
+input DistributionInput {
+    distributionType: DistributionType!
+    channels: [ID!]
+    followers: [ID!]
```
Comment (kaioken) [inline graphql/schemas/Social/message.graphql:37]: @FredPeal we don't say who to send it just to send it to my followers

**Review by kaioken — CHANGES_REQUESTED** [review]

File: graphql/schemas/Social/channels.graphql [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+type SocialChannel {
+    id: ID!
+    name: String!
+    slug: String
+    description: String!
+    entity_namespace: String!
+    entity_id: ID!
+    last_message_id: ID!
+    messages: [Message]
```
Comment (kaioken) [inline graphql/schemas/Social/channels.graphql:14]: @FredPeal [Message!]!

**Review by kaioken — CHANGES_REQUESTED** [review]

File: graphql/schemas/Social/channels.graphql (hunk 1 above)
Comment (FredPeal) [reply graphql/schemas/Social/channels.graphql]: That is a relationship many to many, @hasMany correspond to one has many relationships

File: src/Social/Distribution/Distribution.php [hunk 4]
```diff
@@ -0,0 +1,22 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Kanvas\Social\Distribution;
+
+use Kanvas\Social\Channels\Models\Channel;
+use Kanvas\Social\Messages\Models\Message;
+
+class Distribution
```
Comment (kaioken) [inline src/Social/Distribution/Distribution.php:10]: @FredPeal chsngr the name to the distribution type

**Review by kaioken — CHANGES_REQUESTED** [review]

File: src/Social/Distribution/Distribution.php (hunk 4 above)
Comment (FredPeal) [reply src/Social/Distribution/Distribution.php:10]: Why Distribution type ?

File: app/GraphQL/Social/Mutations/Messages/MessageManagementMutation.php [hunk 5]
```diff
@@ -5,23 +5,40 @@
 namespace App\GraphQL\Social\Mutations\Messages;
 
 use Kanvas\Apps\Models\Apps;
+use Kanvas\Social\Channels\Repositories\ChannelRepository;
+use Kanvas\Social\Distribution\Jobs\SendToChannelJob;
 use Kanvas\Social\Messages\Actions\CreateMessageAction;
+use Kanvas\Social\Messages\Actions\InteractionMessageAction;
 use Kanvas\Social\Messages\DataTransferObject\MessageInput;
+use Kanvas\Social\Messages\Enums\ActivityType;
+use Kanvas\Social\Messages\Enums\DistributionType;
```
Comment (kaioken) [inline app/GraphQL/Social/Mutations/Messages/MessageManagementMutation.php:14]: @FredPeal add enum to the filename

File: app/GraphQL/Social/Mutations/Messages/MessageManagementMutation.php [hunk 6]
```diff
@@ -39,6 +56,43 @@ public function create(mixed $root, array $request)
...
+            $this->distributeChannels($channels, $message, auth()->user());
+        } elseif($distributionType->value == DistributionType::Users->value) {
+            $this->distributeUsers($message);
+        }
+
+        return $message;
+
+    }
+
+    private function distributeChannels(array $channels, Message $message, Users $user): void
```
Comment (kaioken) [inline app/GraphQL/Social/Mutations/Messages/MessageManagementMutation.php:81]: @FredPeal move this logic to a action or service

File: src/Social/Channels/Actions/CreateChannel.php [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+class CreateChannel
+{
+    public function __construct(
+        protected ChannelDto $channelDto
+    ) {
+    }
+
+    public function execute(): Channel
+    {
+        $channel = new Channel();
```
Comment (kaioken) [inline src/Social/Channels/Actions/CreateChannel.php:20]: @FredPeal use a first or create

File: src/Social/Channels/DataTransferObject/Channel.php [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+use Spatie\LaravelData\Data;
+
+class Channel extends Data
+{
+    public function __construct(
+        public Users $users,
+        public string $name = '',
+        public string $description = '',
+        public string $entity_id,
+        public string $entity_namespace,
```
Comment (kaioken) [inline src/Social/Channels/DataTransferObject/Channel.php:17]: @FredPeal look at these comments

File: src/Social/Distribution/Distribution.php (hunk 4 above)
Comment (kaioken) [reply src/Social/Distribution/Distribution.php:10]: cause i dont understand what this is by looking at the name, this is within the domain distribution so, this is referring to feed distribution

**Review by kaioken — CHANGES_REQUESTED** [review]

File: app/GraphQL/Social/Mutations/Channels/ChannelsManagementMutation.php [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        $channel->updateOrFail();
+
+        return $channel;
+    }
+
+    public function deleteChannel(mixed $rootValue, array $request): Channel
+    {
+        $channel = ChannelRepository::getByIdBuilder(auth()->user())
+            ->where('channel_users.roles_id', 1)
+            ->find($request['id']);
```
Comment (kaioken) [inline app/GraphQL/Social/Mutations/Channels/ChannelsManagementMutation.php:51]: @FredPeal firstOrFail or findOrFail?

File: app/GraphQL/Social/Mutations/Channels/ChannelsManagementMutation.php [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+        return $channel;
+    }
+
+    public function attachUserToChannel(mixed $rootValue, array $request): Channel
+    {
+        $channel = ChannelRepository::getById((int)$request['channel_id'], auth()->user());
+        $user = Users::getByIdFromCompany($request['user_id'], auth()->user()->getCurrentCompany());
+
+        $channel->users()->attach($user->id, ['roles_id' => $request['roles_id']]);
```
Comment (kaioken) [inline app/GraphQL/Social/Mutations/Channels/ChannelsManagementMutation.php:63]: @FredPeal we need to verify the role exist add it to todo

File: graphql/schemas/Social/channels.graphql [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+type SocialChannel {
+    id: ID!
+    name: String!
+    slug: String
+    description: String!
+    entity_namespace: String!
+    entity_id: ID!
+    last_message_id: ID!
+    messages: [Message!]! @belongsToMany
+    users: [User] @belongsToMany
```
Comment (kaioken) [inline graphql/schemas/Social/channels.graphql:15]: @FredPeal same behavior for all list [User!]!

File: graphql/schemas/Social/channels.graphql [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+            columns: [
+                "id",
+                "name",
+                "slug",
+                "description",
+                "entity_namespace",
+                "entity_id",
+            ]
+        )
+    ): [SocialChannel] @paginate(
```
Comment (kaioken) [inline graphql/schemas/Social/channels.graphql:49]: @FredPeal we need to add company_id to the channel , so we can filter the content and avoid leaking other company info if they guess the id

File: graphql/schemas/Social/channels.graphql [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+
+extend type Mutation {
+    createSocialChannel(input: SocialChannelInput!): SocialChannel 
+    @field(resolver: "App\\GraphQL\\Social\\Mutations\\Channels\\ChannelsManagementMutation@createChannel")
+    updateSocialChannel(id: ID!, input: SocialChannelInput!): SocialChannel
+    @field(resolver: "App\\GraphQL\\Social\\Mutations\\Channels\\ChannelsManagementMutation@updateChannel")
+    deleteSocialChannel(id: ID!): SocialChannel
+    @field(resolver: "App\\GraphQL\\Social\\Mutations\\Channels\\ChannelsManagementMutation@deleteChannel")
+    attachUserToSocialChannel(channel_id: ID!, user_id: ID!, roles_id: UsersRolesChannel!): SocialChannel
```
Comment (kaioken) [inline graphql/schemas/Social/channels.graphql:63]: @FredPeal you can create a input for all 3 of params and reuse it on detach, to simplify the param

File: graphql/schemas/Social/message.graphql [hunk 14]
```diff
@@ -11,26 +11,67 @@ type Message {
...
+enum InteractionType {
+    LIKE @enum(value: 1)
+    SAVE @enum(value: 2)
+    SHARE @enum(value: 3)
+    REPORT @enum(value: 4)
+}
+
+input DistributionInput {
+    distributionType: DistributionType!
+    channels: [ID!]
```
Comment (kaioken) [inline graphql/schemas/Social/message.graphql:59]: @FredPeal  same thing force that they have to send you the list [ID!]!

File: src/Kanvas/Users/Models/Users.php [hunk 15]
```diff
@@ -283,6 +286,13 @@ public function linkedSources(): HasMany
         return $this->hasMany(UserLinkedSources::class, 'users_id');
     }
 
+    public function channels(): BelongsToMany
```
Comment (kaioken) [inline src/Kanvas/Users/Models/Users.php:289]: @FredPeal move this to a social user trait

File: src/Social/Channels/Actions/CreateChannelAction.php [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+class CreateChannelAction
+{
+    public function __construct(
+        protected ChannelDto $channelDto
+    ) {
+    }
+
+    public function execute(): Channel
+    {
+        $channel = Channel::firstOrCreate([
```
Comment (kaioken) [inline src/Social/Channels/Actions/CreateChannelAction.php:20]: @FredPeal lets add company_id and app_id

File: src/Social/Messages/Actions/DistributeChannelAction.php [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+
+    public function execute()
+    {
+        $channelsDataBase = [];
+        if ($this->channels) {
+            foreach ($this->channels as $channel) {
+                $channelsDataBase[] = ChannelRepository::getById((int)$channel, $this->user);
+            }
+        } else {
+            $channelsDataBase = $user->channels;
```
Comment (kaioken) [inline src/Social/Messages/Actions/DistributeChannelAction.php:29]: @FredPeal look a the error should use $this->user

File: src/Social/Messages/Models/Message.php [hunk 18]
```diff
@@ -113,4 +114,23 @@ public function appModuleMessage(): HasOne
...
+    }
+
+    public function getMyInteraction(): array
+    {
+        $userMessage = UserMessage::where('users_id', auth()->user()->id)
+            ->where('messages_id', $this->id)
+            ->first();
+
+        return [
+            'is_liked' => $userMessage ? $userMessage->is_liked : 0,
```
Comment (kaioken) [inline src/Social/Messages/Models/Message.php:130]: @FredPeal try (int) ($userMessage?->is_liked);

**Review by kaioken — CHANGES_REQUESTED** [review]

**Review by kaioken — APPROVED** [review]

**Pull request merged**